validate.swift

```diff
@@ -1,531 +1,401 @@
-  let pipe = Pipe()
-  let errorPipe = Pipe()
-  let process = dumpPackageProcessAt(directoryURL, outputTo: pipe, errorsTo: errorPipe)
-
-  process.terminationHandler = {
-    process in
-
-    let package: Package
-
-    guard process.terminationStatus == 0 else {
-      let error : PackageError
-      if process.terminationStatus == 15 {
-        error = .dumpTimeout
-      } else {
-        error = .badDump(String(data: errorPipe.fileHandleForReading.readDataToEndOfFile(), encoding: .utf8))
-      }
-      callback(error)
-      return
+func dumpPackage(atURL url: URL, completion: @escaping (Result<Data, ValidatorError>) -> Void) {
+    let pipe = Pipe()
+    let errorPipe = Pipe()
+    let process = dumpPackageProcessAt(url, outputTo: pipe, errorsTo: errorPipe)
+    
+    process.terminationHandler = { process in
+        
+        guard process.terminationStatus == 0 else {
+            let errorDump = String(data: errorPipe.fileHandleForReading.readDataToEndOfFile(), encoding: .utf8)
+            completion(.failure(.badPackageDump(errorDump)))
+            return
+        }
+        
+        completion(.success(pipe.fileHandleForReading.readDataToEndOfFile()))
     }
+    
+    process.launch()
+}
 
+func verifyPackage(url: URL, completion: @escaping (Error?) -> Void) {
     do {
-      package = try decoder.decode(Package.self, from: pipe.fileHandleForReading.readDataToEndOfFile())
+        let packageURL = try getPackageSwiftURL(url: url).get()
+        let localPackageURL = try downloadPackage(url: packageURL).get()
+        
+        dumpPackage(atURL: localPackageURL) { result in
+            switch result {
+            case .failure(let error):
+                completion(error)
+            case .success(let data):
+                let decoder = JSONDecoder()
+                
+                do {
+                    let package = try decoder.decode(Package.self, from: data)
+                    
+                    guard package.products.isEmpty == false else {
+                        completion(ValidatorError.missingProducts)
+                        return
+                    }
+                    
+                    completion(nil)
+                } catch {
+                    completion(error)
+                }
+            }
+        }
     } catch {
-      callback(.decodingError(error))
-      return
-    }
-
-    guard package.products.count > 0 else {
-      callback(.missingProducts)
-      return
+        completion(error)
     }
-    callback(nil)
-  }
-
-  process.launch()
+}
 
+// MARK: - Helpers
 
-  DispatchQueue.main.asyncAfter(deadline: .now() + processTimeout) {
-    if process.isRunning {
-      process.terminate()
+extension Array where Element == URL {
+    
+    func findDuplicates(of url: URL) -> Array<(offset: Int, element: URL)> {
+        let normalise: (URL) -> String = { url in
+            url.absoluteString.lowercased().replacingOccurrences(of: ".git", with: "")
```

Comment:
Maybe only for `hasSuffix(".git")` - because someone will eventually submit "awesome.git.tools" 🤣

Reply:
Made this even more resilient 👍